src/storage/store/list_node_column.cpp

```diff
@@ -0,0 +1,160 @@
+#include "storage/store/list_node_column.h"
+
+#include "storage/copier/list_column_chunk.h"
+#include "storage/store/node_column.h"
+
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace storage {
+
+void ListNodeColumn::scanInternal(transaction::Transaction* transaction,
+    common::ValueVector* nodeIDVector, common::ValueVector* resultVector) {
+    resultVector->resetAuxiliaryBuffer();
+    auto offsetVector = std::make_unique<ValueVector>(LogicalTypeID::INT64);
+    offsetVector->setState(resultVector->state);
+    auto startNodeOffset = nodeIDVector->readNodeOffset(0);
+    auto nodeGroupIdx = getNodeGroupIdxFromNodeOffset(startNodeOffset);
+    scanListOffset(transaction, startNodeOffset, nodeGroupIdx, offsetVector.get());
+    if (resultVector->state->selVector->isUnfiltered()) {
+        scanUnfiltered(transaction, nodeIDVector, offsetVector.get(), resultVector);
+    } else {
+        scanFiltered(transaction, nodeIDVector, offsetVector.get(), resultVector);
+    }
+}
+
+void ListNodeColumn::lookupValue(transaction::Transaction* transaction, common::offset_t nodeOffset,
+    common::ValueVector* resultVector, uint32_t posInVector) {
+    auto listOffset = readListOffsetInStorage(transaction, nodeOffset);
+    auto length = readListOffsetInStorage(transaction, nodeOffset + 1) -
+                  readListOffsetInStorage(transaction, nodeOffset);
+    auto offsetInVector = posInVector == 0 ? 0 : resultVector->getValue<offset_t>(posInVector - 1);
+    resultVector->setValue(posInVector, list_entry_t{offsetInVector, length});
+    ListVector::resizeDataVector(resultVector, offsetInVector + length);
+    dataNodeColumn->scanWithOffsets(transaction, getNodeGroupIdxFromNodeOffset(nodeOffset),
+        listOffset, listOffset + length, ListVector::getDataVector(resultVector), offsetInVector);
+}
+
+page_idx_t ListNodeColumn::append(
+    ColumnChunk* columnChunk, page_idx_t startPageIdx, uint64_t nodeGroupIdx) {
+    auto numPagesFlushed = NodeColumn::append(columnChunk, startPageIdx, nodeGroupIdx);
+    auto dataColumnChunk = reinterpret_cast<ListColumnChunk*>(columnChunk)->getDataColumnChunk();
+    auto numPagesForDataColumn =
+        dataNodeColumn->append(dataColumnChunk, startPageIdx + numPagesFlushed, nodeGroupIdx);
+    return numPagesFlushed + numPagesForDataColumn;
+}
+
+void ListNodeColumn::scanUnfiltered(transaction::Transaction* transaction,
+    ValueVector* nodeIDVector, ValueVector* offsetVector, ValueVector* resultVector) {
+    auto startNodeOffset = nodeIDVector->readNodeOffset(0);
+    auto nodeGroupIdx = getNodeGroupIdxFromNodeOffset(startNodeOffset);
+    auto numValuesToScan = resultVector->state->selVector->selectedSize;
+    offset_t offsetInVector = 0;
```

Comment:
For both `scanUnfiltered` and `scanFiltered`, to make the code logic simpler, we can read out the base offset first through `readOffset`. Then for `getListLength` and `getListOffsetInStorage`, we can pass in the base offset. So those functions can be simplified to not handle reading from storage.